### PR TITLE
Fix subclassOf inferencing

### PR DESCRIPTION
**Pull request by cap10morgan**
Fixes https://github.com/fluree/core/issues/48

### CODE REVIEW

**Comment by cap10morgan** [comment]
Current dependencies on/for this PR:
* main
  * **PR #652** <a href="https://app.graphite.dev/github/pr/fluree/db/652?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #653** <a href="https://app.graphite.dev/github/pr/fluree/db/653?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **PR #661** <a href="https://app.graphite.dev/github/pr/fluree/db/661?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **PR #662** <a href="https://app.graphite.dev/github/pr/fluree/db/662?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/fluree/db/662?utm_source=stack-comment).

File: src/fluree/db/json_ld/vocab.cljc [hunk 1]
```diff
@@ -243,29 +243,62 @@
...
+                    (log/trace "filter-vocab-flakes adding" f "b/c its sid matches a ref flake's o")
+                    (set-conj vfs f))
+
+                  (class-or-prop-flake-sids fs)
+                  (do
+                    (log/trace "filter-vocab-flakes adding" f "b/c its sid matches a type flake's sid")
+                    (set-conj vfs f))
+
+                  :else vfs)))
+            ref-flakes all-flakes)))
```
Comment (dpetran) [inline src/fluree/db/json_ld/vocab.cljc:301]: If we do a `(or ref-flakes #{})` here we can save the overhead of a `nil` check on every `conj`.

File: src/fluree/db/json_ld/vocab.cljc [hunk 2]
```diff
@@ -243,29 +243,62 @@
...
+  [predicate-flakes all-flakes]
+  (let [ref-flakes               (::ref predicate-flakes)
+        class-or-prop-flakes     (::class-or-prop predicate-flakes)
+        ref-flake-os             (->> ref-flakes (map flake/o) set)
+        class-or-prop-flake-sids (->> class-or-prop-flakes (map flake/s) set)
+        predicate-sids           (->> all-flakes (map flake/p) set)
+        set-conj                 (fnil conj #{})]
+    (reduce (fn [vfs f]
+              (let [fs (flake/s f)]
+                (cond
```
Comment (dpetran) [inline src/fluree/db/json_ld/vocab.cljc:284]: Do all of the cases in this `cond` do the same thing?

**Review by dpetran — APPROVED** [review]
I've got some questions about code organization but this looks good.

I think I would prefer a lot less trace logging - when I'm debugging I tend to prefer very granular/bespoke logging for my specific problem , rather than the firehose of every step. But it's not a blocker by any means.

File: src/fluree/db/json_ld/vocab.cljc (hunk 2 above)
Comment (cap10morgan) [reply src/fluree/db/json_ld/vocab.cljc:284]: Yeah except for the logging, which I guess we can drop now. :)

Comment (dpetran) [reply src/fluree/db/json_ld/vocab.cljc:284]: I think it's useful context, so maybe they can live on as comments?

**Comment by cap10morgan** [comment]
> I've got some questions about code organization but this looks good.
> 
> I think I would prefer a lot less trace logging - when I'm debugging I tend to prefer very granular/bespoke logging for my specific problem , rather than the firehose of every step. But it's not a blocker by any means.

This is the same logic that motivated changing logs like these from debug to trace not too long ago. We need to figure out our logging philosophy and overall approach. To me, `trace` implies a firehose. But besides other log levels, there are also ns filters. But regardless of my opinions / understanding, consensus on an overall team-wide approach would be good here. In the meantime I'll probably just start removing these from most of my PRs because I have no idea what we want to end up in the shipped codebase.

**Comment by zonotope** [comment]
> > I've got some questions about code organization but this looks good.
> > I think I would prefer a lot less trace logging - when I'm debugging I tend to prefer very granular/bespoke logging for my specific problem , rather than the firehose of every step. But it's not a blocker by any means.
> 
> This is the same logic that motivated changing logs like these from debug to trace not too long ago. We need to figure out our logging philosophy and overall approach. To me, `trace` implies a firehose. But besides other log levels, there are also ns filters. Regardless of my opinions / understanding, consensus on an overall team-wide approach would be good here. In the meantime I'll probably just start removing these from most of my PRs because I have no idea what we want to end up in the shipped codebase.

I agree with @cap10morgan that TRACE implies a firehose, but I think I also agree with @dpetran than having all of these logs from all of these namespaces makes the logging system much less useful in general. I think moving these logs from DEBUG to TRACE is much better than what we had before in development, but I still find myself editing logback-test.xml locally to set the log level to DEBUG when I run my tests because the default log level when we run our tests is still TRACE. 

When I have a test failure and get a wall of irrelevant logs in my terminal, it's impossible for me to find what the actual cause of the failure was. A lot of the time there would be so many logs printed out that the scrollback buffer for my terminal fills up and I can't even see the actual test failure. I usually have to edit logback-test.xml at that point to set the log level higher and then add my own logs to pinpoint on the cause of the problem. I had been meaning to bring this up, but just haven't gotten around to it yet.

I'm fine with keeping all of these "stepping through the code" logs at trace level, but I think we should change the default log level for tests to DEBUG like we have in development. I don't think TRACE should be enabled by default anywhere, but it might be useful in some situation to be able to trace the code execution step by step, so we should still have the option.

So, in short, my preference would be that the default log level is DEBUG for test and development profiles, and INFO for production, and that TRACE isn't enabled by default anywhere, but we still have the option to enable it through the logging config.
